src/api.cpp

```diff
@@ -4590,195 +4591,47 @@ cl_int CLVK_API_CALL clGetImageInfo(cl_mem image, cl_image_info param_name,
-};
-
-static void get_component_mappings_for_channel_order(
-    cl_channel_order order, VkComponentMapping* components_sampled,
-    VkComponentMapping* components_storage) {
-    if (order == CL_LUMINANCE) {
-        *components_sampled = {VK_COMPONENT_SWIZZLE_R, VK_COMPONENT_SWIZZLE_R,
-                               VK_COMPONENT_SWIZZLE_R, VK_COMPONENT_SWIZZLE_A};
-    } else if (order == CL_INTENSITY) {
-        *components_sampled = {VK_COMPONENT_SWIZZLE_R, VK_COMPONENT_SWIZZLE_R,
-                               VK_COMPONENT_SWIZZLE_R, VK_COMPONENT_SWIZZLE_R};
-    } else {
-        *components_sampled = {
-            VK_COMPONENT_SWIZZLE_IDENTITY, VK_COMPONENT_SWIZZLE_IDENTITY,
-            VK_COMPONENT_SWIZZLE_IDENTITY, VK_COMPONENT_SWIZZLE_IDENTITY};
-    }
-
-    *components_storage = {
-        VK_COMPONENT_SWIZZLE_IDENTITY, VK_COMPONENT_SWIZZLE_IDENTITY,
-        VK_COMPONENT_SWIZZLE_IDENTITY, VK_COMPONENT_SWIZZLE_IDENTITY};
+bool operator!=(const VkComponentMapping& lhs, const VkComponentMapping& rhs) {
+    return lhs.r != rhs.r || lhs.g != rhs.g || lhs.b != rhs.b || lhs.a != rhs.a;
 }
 
-bool cl_image_format_to_vulkan_format(const cl_image_format& clformat,
-                                      VkFormat* format,
-                                      VkComponentMapping* components_sampled,
-                                      VkComponentMapping* components_storage) {
-    auto m = gFormatMaps.find(clformat);
-    bool success = false;
-
-    if (m != gFormatMaps.end()) {
-        *format = (*m).second.vkfmt;
-        success = true;
+static bool is_image_format_supported(
+    VkPhysicalDevice& pdev, VkFormat format, cl_mem_object_type image_type,
+    const VkFormatFeatureFlags& required_format_feature_flags,
+    cl_channel_order image_channel_order) {
+    VkFormatProperties properties;
+    vkGetPhysicalDeviceFormatProperties(pdev, format, &properties);
+
+    cvk_debug("Vulkan format %d:", format);
+    cvk_debug(
+        "  linear : %s",
+        vulkan_format_features_string(properties.linearTilingFeatures).c_str());
+    cvk_debug("  optimal: %s",
+              vulkan_format_features_string(properties.optimalTilingFeatures)
+                  .c_str());
+    cvk_debug("  buffer : %s",
+              vulkan_format_features_string(properties.bufferFeatures).c_str());
+
+    cvk_debug(
+        "Required format features %s",
+        vulkan_format_features_string(required_format_feature_flags).c_str());
+    VkFormatFeatureFlags features;
+    if (image_type == CL_MEM_OBJECT_IMAGE1D_BUFFER) {
+        features = properties.bufferFeatures;
+    } else {
+        // TODO support linear tiling
+        features = properties.optimalTilingFeatures;
+    }
+    if ((features & required_format_feature_flags) ==
+        required_format_feature_flags) {
+        if ((image_channel_order == CL_LUMINANCE ||
+             image_channel_order == CL_INTENSITY) &&
+            (image_type == CL_MEM_OBJECT_IMAGE1D_BUFFER)) {
```

Comment:
It still doesn't capture the reason but at least the intent is now very clear. That's a step up.

Reply:
I don't know more than `IMAGE1D_BUFFER` and `CL_LUMINANCE` or `CL_INTENSITY` are not supported at the moment. I could add something like that in a comment, but I'm not sure that's adding anything.

Reply:
Same here. I don't remember why. I'd apparently used a condition based on swizzles in my initial patch but neither of us could understand this. Don't sweat it, we'll figure it out as part of #625.